File: source/usfocuscommands.ts
// unit usfocuscommands

import { int } from "./common"
import { KfCommand } from "./KfCommand"

// TODO: Remove
const usdomain: any = {}
type TWinControl = any
type TForm = any
type TModelChange = any
class TControl {Text: string; selStart: number; selLength: number}
class TCustomEdit extends TControl {}
class TQuickFillComboBox extends TControl {}

export interface TSelectionInformation {
    text: string
    selStart: int
    selLength: int
}

function RecordSelectionInformation(selectionInformation: TSelectionInformation, control: TControl): void {
    if (control instanceof TCustomEdit) {
        selectionInformation.text = control.Text
        selectionInformation.selStart = control.selStart
        selectionInformation.selLength = control.selLength
    } else if (control instanceof TQuickFillComboBox) {
        selectionInformation.text = control.Text
        selectionInformation.selStart = control.selStart
        selectionInformation.selLength = control.selLength
    } else {
        // error
        throw new Error("RecordSelectionInformation: unsupported control class")
    }
}

function RestoreSelectionInformation(selectionInformation: TSelectionInformation, control: TControl): void {
    if (control instanceof TCustomEdit) {
        control.Text = selectionInformation.text
        control.selStart = selectionInformation.selStart
        control.selLength = selectionInformation.selLength
    } else if (control instanceof TQuickFillComboBox) {
        control.Text = selectionInformation.text
        control.selStart = selectionInformation.selStart
        control.selLength = selectionInformation.selLength
    } else {
        // error
        throw new Error("RestoreSelectionInformation: unsupported control class")
    }
}

export class TFocusShiftAction {
    control: TWinControl
    form: TForm
    modelChange: TModelChange
    
    // createWithFormControlModelChange
    constructor(form: TForm, control: TWinControl, modelChange: TModelChange) {
        this.form = form
        this.control = control
        this.modelChange = modelChange
    }
    
    // param ignored
    recordState(selectionInformation: TSelectionInformation): void {
        // subclasses should override
    }
    
    // param ignored
    restoreStateWithFocus(): void {
        // form.Active := true;
        this.control.SetFocus()
        // subclasses may override
    }
    
    // assume focus shifted is handled by system the first time
    // assume do must be called before other component is focused to if it is off the form
    doShift(): void {
        if (this.modelChange !== null) {
            // subclasses may override
            this.modelChange.doChange()
        }
    }
    
    undoShift(): void {
        if (this.modelChange !== null) {
            // subclasses may override
            this.modelChange.undoChange()
        }
    }
    
    redoShift(): void {
        if (this.modelChange !== null) {
            this.modelChange.redoChange()
        }
    }
    
}

export class TFocusExitAction extends TFocusShiftAction {
    
    doShift(): void {
        super.doShift()
    }
    
    // exit moves focus on undo
    undoShift(): void {
        this.restoreStateWithFocus()
        // need to restore selectionInformation in form
        super.undoShift()
    }
    
    redoShift(): void {
        super.redoShift()
    }
    
}

export class TFocusEnterAction extends TFocusShiftAction {
    
    doShift(): void {
        super.doShift()
    }
    
    undoShift(): void {
        super.undoShift()
    }
    
    // enter moves focus on redo
    redoShift(): void {
        this.restoreStateWithFocus()
        super.redoShift()
    }
    
}

export class TTextFocusExitAction extends TFocusExitAction {
    selectionInformation: TSelectionInformation
    
    recordState(selectionInformation: TSelectionInformation): void {
        super.recordState(selectionInformation)
        this.selectionInformation = selectionInformation
        // RecordSelectionInformation(selectionInformation, control)
    }
    
    restoreStateWithFocus(): void {
        super.restoreStateWithFocus()
        RestoreSelectionInformation(this.selectionInformation, this.control)
    }
    
}

export class TTextFocusEnterAction extends TFocusEnterAction {
    
}

export class TFocusShiftCommand extends KfCommand {
    focusExitAction: TFocusExitAction
    focusEnterAction: TFocusEnterAction
    
    constructor(focusExitAction: TFocusExitAction) {
        super()
        this.focusExitAction = focusExitAction
        //self.focusExitAction.recordState;
        // responsibility of component being focused to call setFocusEnterAction
    }
    
    setFocusEnterAction(focusEnterAction: TFocusEnterAction): void {
        let selectionInformation: TSelectionInformation = {text: "", selStart: 0, selLength: 0}
        
        this.focusEnterAction = focusEnterAction
        // param ignored
        this.focusEnterAction.recordState(selectionInformation)
    }
    
    doCommand(): void {
        super.doCommand()
        usdomain.domain.beginUpdate()
        this.focusExitAction.doShift()
        if (this.focusEnterAction !== null) {
            this.focusEnterAction.doShift()
        }
        usdomain.domain.endUpdate()
    }
    
    undoCommand(): void {
        super.undoCommand()
        usdomain.domain.beginUpdate()
        if (this.focusEnterAction !== null) {
            this.focusEnterAction.undoShift()
        }
        this.focusExitAction.undoShift()
        usdomain.domain.endUpdate()
    }
    
    redoCommand(): void {
        super.doCommand()
        usdomain.domain.beginUpdate()
        this.focusExitAction.redoShift()
        if (this.focusEnterAction !== null) {
            this.focusEnterAction.redoShift()
        }
        usdomain.domain.endUpdate()
    }
    
    description(): string {
        let result = ""
        result = "editor focus shift"
        return result
    }
    
}

File: converted_source/used/usfocuscommands.pas
unit usfocuscommands;

interface

uses Windows, Graphics, Classes, Forms, Controls, StdCtrls, UCommand, USWorld, 
	USModelChanges;

type

  TSelectionInformation = record
 		text: string;
		selStart, selLength: integer;
    end;

  procedure RecordSelectionInformation(var selectionInformation: TSelectionInformation; control: TControl);
  procedure RestoreSelectionInformation(const selectionInformation: TSelectionInformation; control: TControl);

type

  TFocusShiftAction = class(TObject)
    control: TWinControl;
    form: TForm;
    modelChange: TModelChange;
    constructor createWithFormControlModelChange(form: TForm; control: TWinControl; modelChange: TModelChange); virtual;
    destructor destroy; override;
    procedure recordState(const selectionInformation: TSelectionInformation); virtual;  // param ignored
    procedure restoreStateWithFocus; virtual;
    procedure doShift; virtual;
    procedure undoShift; virtual;
    procedure redoShift; virtual;
    end;

	TFocusExitAction = class(TFocusShiftAction)
    procedure doShift; override;
    procedure undoShift; override;
    procedure redoShift; override;
    end;

	TFocusEnterAction = class(TFocusShiftAction)
    procedure doShift; override;
    procedure undoShift; override;
    procedure redoShift; override;
    end;

  TTextFocusExitAction = class(TFocusExitAction)
 		selectionInformation: TSelectionInformation;
    procedure recordState(const selectionInformation: TSelectionInformation); override;
    procedure restoreStateWithFocus; override;
    end;

  TTextFocusEnterAction = class(TFocusEnterAction)
    end;

	TFocusShiftCommand = class(KfCommand)
    focusExitAction: TFocusExitAction;
    focusEnterAction: TFocusEnterAction;
    constructor createWithFocusExitAction(focusExitAction: TFocusExitAction);
    procedure setFocusEnterAction(focusEnterAction: TFocusEnterAction);
    destructor destroy; override;
 		procedure doCommand; override;
		procedure undoCommand; override;
		procedure redoCommand; override;
 		function description: string; override;
    end;

implementation

uses SysUtils, QuickFillComboBox, USDomain;

procedure RecordSelectionInformation(var selectionInformation: TSelectionInformation; control: TControl);
	begin
  if control is TCustomEdit then
    begin
    selectionInformation.text := (control as TCustomEdit).text;
  	selectionInformation.selStart := (control as TCustomEdit).selStart;
  	selectionInformation.selLength := (control as TCustomEdit).selLength;
    end
  else if control is TQuickFillComboBox then
    begin
    selectionInformation.text := (control as TQuickFillComboBox).text;
  	selectionInformation.selStart := (control as TQuickFillComboBox).selStart;
  	selectionInformation.selLength := (control as TQuickFillComboBox).selLength;
    end
  else
    begin
    // error
    raise Exception.create('RecordSelectionInformation: unsupported control class');
    end;
  end;

procedure RestoreSelectionInformation(const selectionInformation: TSelectionInformation; control: TControl);
	begin
  if control is TCustomEdit then
    begin
    (control as TCustomEdit).text := selectionInformation.text;
  	(control as TCustomEdit).selStart := selectionInformation.selStart;
  	(control as TCustomEdit).selLength := selectionInformation.selLength;
    end
  else if control is TQuickFillComboBox then
    begin
    (control as TQuickFillComboBox).text := selectionInformation.text;
  	(control as TQuickFillComboBox).selStart := selectionInformation.selStart;
  	(control as TQuickFillComboBox).selLength := selectionInformation.selLength;
    end
  else
    begin
    // error
    raise Exception.create('RestoreSelectionInformation: unsupported control class');
    end;
  end;

///////////////////////////// TFocusShiftAction //////////////////////////

constructor TFocusShiftAction.createWithFormControlModelChange(form: TForm; control: TWinControl; modelChange: TModelChange);
	begin
  self.form := form;
  self.control := control;
  self.modelChange := modelChange;
  end;

destructor TFocusShiftAction.destroy;
  begin
  modelChange.free;
  inherited destroy;
  end;

procedure TFocusShiftAction.recordState(const selectionInformation: TSelectionInformation); // param ignored
	begin
  // subclasses should override
  end;

procedure TFocusShiftAction.restoreStateWithFocus;
	begin
  // form.Active := true;
  control.setFocus;
  // subclasses may override
  end;

// assume focus shifted is handled by system the first time
// assume do must be called before other component is focused to if it is off the form
procedure TFocusShiftAction.doShift;
	begin
  // subclasses may override
  if modelChange <> nil then
  	modelChange.doChange;
  end;

procedure TFocusShiftAction.undoShift;
	begin
  // subclasses may override
  if modelChange <> nil then
    modelChange.undoChange;
  end;

procedure TFocusShiftAction.redoShift;
	begin
 	if modelChange <> nil then
    modelChange.redoChange;
  end;

//////////////////////////// TFocusExitAction ////////////////////////////////

procedure TFocusExitAction.doShift;
	begin
  inherited doShift;
  end;

// exit moves focus on undo
procedure TFocusExitAction.undoShift;
	begin
  self.restoreStateWithFocus;
  // need to restore selectionInformation in form
 	inherited undoShift;
  end;

procedure TFocusExitAction.redoShift;
	begin
 	inherited redoShift;
  end;

//////////////////////////// TFocusEnterAction ////////////////////////////////

procedure TFocusEnterAction.doShift;
	begin
  inherited doShift;
  end;

procedure TFocusEnterAction.undoShift;
	begin
 	inherited undoShift;
  end;

// enter moves focus on redo
procedure TFocusEnterAction.redoShift;
	begin
  self.restoreStateWithFocus;
 	inherited redoShift;
  end;

//////////////////////////// TTextFocusExitAction ////////////////////////////////

procedure TTextFocusExitAction.recordState(const selectionInformation: TSelectionInformation);
  begin
  inherited recordState(selectionInformation);
  self.selectionInformation := selectionInformation;
  //RecordSelectionInformation(selectionInformation, control);
  end;

procedure TTextFocusExitAction.restoreStateWithFocus;
  begin
  inherited restoreStateWithFocus;
  RestoreSelectionInformation(selectionInformation, control);
  end;

//////////////////////////// TTextFocusEnterAction ////////////////////////////////

//////////////////////////// TFocusShiftCommand ////////////////////////////////

constructor TFocusShiftCommand.createWithFocusExitAction(focusExitAction: TFocusExitAction);
  begin
  self.create;
  self.focusExitAction := focusExitAction;
  //self.focusExitAction.recordState;
  // responsibility of component being focused to call setFocusEnterAction
  end;

procedure TFocusShiftCommand.setFocusEnterAction(focusEnterAction: TFocusEnterAction);
  var selectionInformation: TSelectionInformation;
  begin
  self.focusEnterAction := focusEnterAction;
  self.FocusEnterAction.recordState(selectionInformation); // param ignored
  end;

destructor TFocusShiftCommand.destroy;
  begin
  focusExitAction.free;
  focusEnterAction.free;
  inherited destroy;
  end;

procedure TFocusShiftCommand.doCommand;
  begin
  inherited doCommand;
  domain.beginUpdate;
  focusExitAction.doShift;
  if focusEnterAction <> nil then
  	focusEnterAction.doShift;
  domain.endUpdate;
  end;

procedure TFocusShiftCommand.undoCommand;
  begin
  inherited undoCommand;
  domain.beginUpdate;
  if focusEnterAction <> nil then
  	focusEnterAction.undoShift;
  focusExitAction.undoShift;
  domain.endUpdate;
  end;

procedure TFocusShiftCommand.redoCommand;
  begin
  inherited doCommand;
  domain.beginUpdate;
  focusExitAction.redoShift;
  if focusEnterAction <> nil then
  	focusEnterAction.redoShift;
  domain.endUpdate;
  end;

function TFocusShiftCommand.description: string;
	begin
  result := 'editor focus shift';
  end;

end.
